### PR TITLE
FileSearch: Let callers provide a filter predicate

### PR DESCRIPTION
**Pull request by JosJuice**
This will let me make a part of PR #5573 simpler and faster (the `DiscIO::ShouldHideFromGameList` filtering). It also decreases the amount of platform-specific code in FileSearch.cpp.

### CODE REVIEW

File: Source/Core/Common/FileSearch.cpp [hunk 1]
```diff
@@ -45,57 +45,17 @@ FileSearchWithTest(const std::vector<std::string>& directories, bool recursive,
...
-    });
-  };
-
   std::vector<std::string> result;
   auto add_filtered = [&](const fs::directory_entry& entry) {
-    auto& path = entry.path();
-    if (accept_all || (ext_matches(path) && !fs::is_directory(path)))
-      result.emplace_back(path.u8string());
+    const std::string u8_path = entry.path().u8string();
+    if (!filter || filter(u8_path, fs::is_directory(entry.path())))
```
Comment (shuffle2) [inline Source/Core/Common/FileSearch.cpp]: do not call `is_directory` when not required, it is slow.

File: Source/Core/Common/FileSearch.cpp [hunk 2]
```diff
@@ -129,4 +89,22 @@ std::vector<std::string> DoFileSearch(const std::vector<std::string>& directorie
...
 
+std::vector<std::string>
+DoFileSearch(const std::vector<std::string>& directories, bool recursive,
+             const std::vector<std::string>& exts,
+             std::function<bool(const std::string& path, bool is_directory)> predicate)
+{
+  return DoFileSearch(directories, recursive, [&](const std::string& path, bool is_directory) {
+    if (is_directory)
+      return false;
+    std::string path_copy = path;
```
Comment (shuffle2) [inline Source/Core/Common/FileSearch.cpp]: this does a copy which i explicitly avoided. it is very slow.

File: Source/Core/Common/FileSearch.h [hunk 3]
```diff
@@ -4,14 +4,17 @@
...
-// Callers can pass empty "exts" to indicate they want all files + directories in results
-// Otherwise, only files matching the extensions are returned
-std::vector<std::string> DoFileSearch(const std::vector<std::string>& directories,
-                                      const std::vector<std::string>& exts = {},
-                                      bool recursive = false);
+std::vector<std::string>
+DoFileSearch(const std::vector<std::string>& directories, bool recursive = false,
+             std::function<bool(const std::string& path, bool is_directory)> predicate = {});
+std::vector<std::string>
+DoFileSearch(const std::vector<std::string>& directories, bool recursive,
```
Comment (shuffle2) [inline Source/Core/Common/FileSearch.h]: If you move recursive to the second arg, there should be many more callsites that need to change...

**Comment by JosJuice** [comment]
Closing for now due to the performance problems.

File: Source/Core/Common/FileSearch.cpp [hunk 4]
```diff
@@ -45,57 +45,17 @@ FileSearchWithTest(const std::vector<std::string>& directories, bool recursive,
...
-                      ext.c_str()) == 0;
-    });
-  };
-
   std::vector<std::string> result;
   auto add_filtered = [&](const fs::directory_entry& entry) {
-    auto& path = entry.path();
-    if (accept_all || (ext_matches(path) && !fs::is_directory(path)))
-      result.emplace_back(path.u8string());
+    const std::string u8_path = entry.path().u8string();
```
Comment (JosJuice) [inline Source/Core/Common/FileSearch.cpp]: This also does a copy, but we more or less need this for this PR's design to work... So I suppose this PR's design is unworkable if we care about performance.

**Pull request closed without merging**